traits/trait_set_object.py

```diff
@@ -12,247 +12,425 @@
-            if items_event is None and hasattr(self, "trait"):
-                items_event = self.trait.items_event()
-            object.trait_items_event(name, event, items_event)
+
+        if self.validator is None:
+            return value
+        else:
+            return self.validator(self, value)
+
+    def notify(self, removed, added):
+        """ Call all notifiers.
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(removed, added)
+
+        Any return values are ignored.
+
+        Parameters
+        ----------
+        removed : set
+            The items to be removed.
+        added : set
+            The new items being added to the set.
+        """
+        for notifier in self.notifiers:
+            notifier(removed, added)
+
+    # ------------------------------------------------------------------------
+    # set interface
+    # ------------------------------------------------------------------------
 
     def __deepcopy__(self, memo):
+        """ Perform a deepcopy operation.
+
+        Notifiers are transient and should not be copied.
+        """
         id_self = id(self)
         if id_self in memo:
             return memo[id_self]
 
-        memo[id_self] = result = TraitSetObject(
-            self.trait,
-            lambda: None,
-            self.name,
+        # notifiers are transient and should not be copied
+        memo[id_self] = result = TraitSet(
             [copy.deepcopy(x, memo) for x in self],
+            validator=copy.deepcopy(self.validator, memo),
+            notifiers=[],
         )
 
         return result
 
-    def update(self, value):
-        if not hasattr(self, "trait"):
-            set.update(self, value)
-            return
+    def __getstate__(self):
+        """ Get the state of the object for serialization.
 
-        try:
-            if not isinstance(value, set):
-                value = set(value)
-            added = value.difference(self)
-            if len(added) > 0:
-                object = self.object()
-                validate = self.trait.item_trait.handler.validate
-                if validate is not None:
-                    name = self.name
-                    added = set(
-                        [validate(object, name, item) for item in added]
-                    )
-
-                set.update(self, added)
-
-                if self.name_items is not None:
-                    self._send_trait_items_event(
-                        self.name_items, TraitSetEvent(None, added)
-                    )
-        except TraitError as excp:
-            excp.set_prefix("Each element of the")
-            raise excp
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        # notifiers are transient and should not be serialized
+        result.pop("notifiers", None)
```

Comment:
`del result["notifiers"]` would be simpler. (If `notifiers` isn't in the dictionary, that's an indication of a bug in our logic somewhere, so I don't think we need to guard against that case.)